core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java

```diff
@@ -885,14 +886,22 @@ public TrimResult trimFields(
 
   /**
    * Variant of {@link #trimFields(RelNode, ImmutableBitSet, Set)} for
-   * {@link org.apache.calcite.rel.core.SetOp} (including UNION and UNION ALL).
+   * {@link org.apache.calcite.rel.core.SetOp} (Only UNION ALL is supported).
    */
   public TrimResult trimFields(
       SetOp setOp,
       ImmutableBitSet fieldsUsed,
       Set<RelDataTypeField> extraFields) {
     final RelDataType rowType = setOp.getRowType();
     final int fieldCount = rowType.getFieldCount();
+
+    // Trim fields only for UNION ALL.
+    // Operator UNION/INTERSECT/INTERSECT ALL/EXCEPT/EXCEPT ALL
+    // will check duplication. We can not trim fields on them.
+    if (!(setOp.kind == SqlKind.UNION && setOp.all)) {
+      return result(setOp, Mappings.createIdentity(fieldCount));
```

Comment:
Does `INTERSECT ALL` and `EXCEPT ALL` check the duplicates ?

Reply:
> Does `INTERSECT ALL` and `EXCEPT ALL` check the duplicates ?

Seems they don't. This is in the ori JIRA, should I remove them from the can not trim list?

Reply:
You may need to check it with different db engines, here https://www.db-fiddle.com/

Reply:
PG:
![image](https://user-images.githubusercontent.com/15643702/91943901-df128480-ed2f-11ea-8afb-2f240d57dd37.png)

![image](https://user-images.githubusercontent.com/15643702/91943944-ee91cd80-ed2f-11ea-861c-41d632528815.png)

Reply:
PG does have the difference, and MySQL seems don't support intersect all/except all.

Reply:
And what is your conclusion there ？

Reply:
> And what is your conclusion there ？

We can not trim fields on the operator INTERSECT ALL/EXCEPT ALL

Reply:
My test shows that INTERSECT/EXCEPT ALL does not distinct values, so we can trim fields for them.
Here is my test query:
```sql
create table a (
  a1 int,
  a2 int
);

insert into a values (1, 2);
insert into a values (3, 4);
insert into a values (3, 4);
insert into a values (1, 2);
insert into a values (5, 6);
insert into a values (5, 6);

create table b (
  b1 int,
  b2 int
);

insert into b values (1, 2);
insert into b values (3, 4);
insert into b values (3, 4);
```

```sql
select * from a except all select * from b;
```

Reply:
INTERSECT/EXCEPT ALL does not distinct values, but we still can not trim fields for them.
Take the pic below, for example, q1 is the trimmed form of q2, and we can the result in the pic, they are different:
q1:
```
select name from
(select name from emp0
intersect all
select name from emp1)
```

q2:
```
select name from
(select name, deptno from emp0
intersect all
select name, deptno from emp1)
```

![image](https://user-images.githubusercontent.com/15643702/92077436-02503900-edef-11ea-966a-cf5095167d02.png)

Reply:
Hmm, i think i got your point: any set operations that have comparison between branches can not trim fields.